### PR TITLE
James-1914 Mix jmap imap

### PR DESCRIPTION
**Pull request by strokyl**
Please ignore commit about James-1785 (there will be merged with [an other PR](https://github.com/linagora/james-project/pull/570))

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java [hunk 1]
```diff
@@ -45,6 +45,14 @@ public boolean userReceivedMessageInMailbox(String user, String password, String
             .contains("OK FETCH completed");
     }
 
+    public boolean userGetNotifiedForNewMessagesWhenSelectingMailbox(String user, String password, int numOfNewMessage, String mailboxName) throws IOException {
+        imapClient.login(user, password);
+        imapClient.select(mailboxName);
+        System.out.println(imapClient.getReplyString());
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: println ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature [hunk 2]
```diff
@@ -19,4 +19,9 @@ Feature: IMAP compatibility of JMAP setMessages method used to update mailboxes
     Given the user has a message "m1" in "source" mailbox with subject "My awesome subject", content "This is the content"
     When the user copy "m1" from mailbox "source" to mailbox "mailbox"
     Then the user has a IMAP message in mailbox "mailbox"
-    And the user has a IMAP message in mailbox "source"
\ No newline at end of file
+    And the user has a IMAP message in mailbox "source"
+
+  Scenario: If a message is moved byJMAP, IMAP connexion will be notified
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: missing space

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature [hunk 3]
```diff
@@ -21,7 +21,13 @@ Feature: IMAP compatibility of JMAP setMessages method used to update mailboxes
...
 
-  Scenario: If a message is moved byJMAP, IMAP connexion will be notified
+  Scenario: If a message is moved by JMAP, IMAP client will be notified when selecting mailbox
     Given the user has a message "m1" in "source" mailbox with subject "My awesome subject", content "This is the content"
     When the user move "m1" to mailbox "mailbox"
-    Then the user has a IMAP notification about 1 new message on mailbox "mailbox"
\ No newline at end of file
+    Then the user has a IMAP notification about 1 new message when selecting mailbox "mailbox"
+
+  Scenario: If a message is moved by JMAP, IMAP client that have selected the destination mailbow will be notified
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: mailbow?

Comment (strokyl) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: Oups corrected !

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature (hunk 2 above)
Comment (strokyl) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: Done

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java (hunk 1 above)
Comment (strokyl) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: Oups done

**Comment by chibenwa** [comment]
My fault, missing one depandancy for Cassandra...

```
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840] [INFO] -------------------------------------------------------------
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840] [ERROR] COMPILATION ERROR : 
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840] [INFO] -------------------------------------------------------------
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840] [ERROR] /james-project/server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java:[29,38] package org.apache.james.mailets.utils does not exist
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840] [ERROR] /james-project/server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java:[40,31] cannot find symbol
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840]   symbol:   class IMAPMessageReader
[Building Java 8 mix-jmap-imap-714b3dc4-020e-45b2-8b71-a331c023f840]   location: class org.apache.james.jmap.methods.integration.cucumber.ImapStepdefs
```

**Comment by chibenwa** [comment]
I was actually thinking to IMAP IDLE command...

**Comment by chibenwa** [comment]
Ok, so after checks commons-net do not support IDLE command. It can not be easily extended.

Can be a contribution :-p I checked there JIRA, no issue on this.

However I do think this is out of Sprint ;-)

**Comment by chibenwa** [comment]
Added a test : https://github.com/chibenwa/james-project/commit/1ddc1371808f9b87d1706feec95690141a195bf6

With these two commits, consider you have my :+1:

**Comment by strokyl** [comment]
Could not find any IDLE command on IMAP rfc. I think you wanted to mean
"NOOP"

Le lun. 23 janv. 2017 02:50, Tellier Benoit <notifications@github.com> a
écrit :

**Comment by chibenwa** [comment]
https://tools.ietf.org/html/rfc2177

**Comment by strokyl** [comment]
After investigation, it's not possible to add test for IDLE without contributing to the under library "org.apache.commons.net.imap"

**Comment by strokyl** [comment]
test this please

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java [hunk 4]
```diff
@@ -75,7 +75,7 @@ public String readFirstMessageInInbox(String user, String password) throws IOExc
     public boolean userGetNotifiedForNewMessages(int numOfMessage) throws IOException {
         imapClient.noop();
 
-        return imapClient.getReplyString().contains(numOfMessage + " RECENT");
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: Not a problem on my laptop

Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: Anyway, we did not added enough details on this....

The reply string was : 

```
 * 1 RECENT
 * 1 EXISTS
TAG OK noop completed.
```

And : 

```
 * 1 EXPUNGED
TAG OK noop completed.
```

I performed these checks more in depth... As well as rebase on master.

**Comment by chibenwa** [comment]
https://github.com/linagora/james-project/pull/580

**Pull request closed without merging**